File: lc/coreprep/implementation/practice_iterations/dp_iterations_5/knapsack_brute.cpp
#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <list>
#include <queue>
#include <random>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;

int next_random(int low, int high) {
  random_device rd;
  mt19937 engine(rd());
  uniform_int_distribution<int> dist(low, high);

  return dist(engine);
}

int knapsack(const vector<int> &profits, const vector<int> &weights,
             int capacity, int curr_idx) {
  if (curr_idx >= weights.size() || capacity == 0) {
    return 0;
  }

  int profit1 = 0;

  if (weights[curr_idx] <= capacity) {
    profit1 =
        profits[curr_idx] +
        knapsack(profits, weights, capacity - weights[curr_idx], curr_idx);
  }

  int profit2 = knapsack(profits, weights, capacity, curr_idx + 1);

  return max(profit1, profit2);
}

// O(2n) / O(n)
int knapsack(const vector<int> &profits, const vector<int> &weights,
             int capacity) {
  if (weights.empty() || capacity == 0) {
    return 0;
  }

  return knapsack(profits, weights, capacity, 0);
}

int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);

  int tt, n, capacity;
  cin >> tt;

  while (tt--) {
    cin >> n >> capacity;

    vector<int> profits(n);
    for (int i = 0; i < n; i++) {
      cin >> profits[i];
    }

    vector<int> weights(n);
    for (int i = 0; i < n; i++) {
      cin >> weights[i];
    }

    cout << knapsack(profits, weights, capacity) << "\n";
  }

  return 0;
}

File: lc/coreprep/implementation/dp_reference_implementation/KnapsackBrute.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class KnapsackBrute {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();
        int c = in.nextInt();

        int[] profits = new int[n];
        for (int i = 0; i < n; i++) {
          profits[i] = in.nextInt();
        }

        int[] weights = new int[n];
        for (int i = 0; i < n; i++) {
          weights[i] = in.nextInt();
        }

        System.out.println(knapsack(profits, weights, c));
      }
    }
  }

  // O(2n) / O(n)
  public static int knapsack(int[] profits, int[] weights, int capacity) {
    if (weights.length == 0 || capacity == 0) {
      return 0;
    }

    return knapsack(profits, weights, capacity, 0);
  }

  private static int knapsack(int[] profits, int[] weights, int capacity,
                              int currIdx) {
    if (currIdx >= weights.length) {
      return 0;
    }

    int profit1 = 0;
    if (weights[currIdx] <= capacity) {
      profit1 =
          profits[currIdx] +
          knapsack(profits, weights, capacity - weights[currIdx], currIdx);
    }

    int profit2 = knapsack(profits, weights, capacity, currIdx + 1);

    return Math.max(profit1, profit2);
  }
}